### ISSUE TITLE
mbedtls-security-advisory-2021-07-1, -2021-07-2. CVE-2021-44732. Update to Mbed TLS 2.28.0 

### ISSUE THREAD

**Issue by mkitti** [JuliaLang/julia#43677]
There have been several references to these security advisories for mbedtls in the issues / PRs in past few months. I'm not sure if they have received the needed attention since security advisory was not obvious in the title and there are no replies on the previous posts.

To make the issue clearer, I have produced the text of the advisories below.

# Local side channel attack on RSA
https://tls.mbed.org/tech-updates/security-advisories/mbedtls-security-advisory-2021-07-1

## Vulnerability
The Montgomery curves Curve25519 and Curve448, also known as x25519 and x448 when used for Diffie-Hellman, were designed to minimize the number of checks an implementation needs to do for secure use.

In particular, validity of the peer's public key needs not be checked, as long as the underlying multi-precision (bignum) arithmetic is constant-time. This is not the case in Mbed TLS, but validity checks were still skipped, so an attacker could exploit special inputs (low-order points) in order to cause variations in timing and memory access patterns that would in turn leak information about the private key.

## Impact
An attacker with access to precise enough timing and memory access information (for example, able to execute arbitrary code and sharing a memory cache with the victim) can recover the private keys used in static Diffie-Hellman with x25519 and x448.

## Resolution
Affected users will want to upgrade to Mbed TLS 3.0.0, 2.27.0 or 2.16.11 depending on the branch they're currently using.

# Local side channel attack on static Diffie-Hellman with Montgomery curves
https://tls.mbed.org/tech-updates/security-advisories/mbedtls-security-advisory-2021-07-2

## Vulnerability
The modular exponentiation operation in RSA uses a sliding window algorithm, with a memory access pattern that depends on the bits of the secret key.

Exponent blinding is used as a counter-measure: it prevents an attacker from correlating informations gathered on successive operation, but researchers found a way to recover enough information by observing a single operation, therefore by-passing this counter-measure.

## Impact
An attacker with access to precise enough timing and memory access information (typically an untrusted operating system attacking a secure enclave such as SGX or the TrustZone secure world) can recover the private keys used in RSA.

## Resolution
Affected users will want to upgrade to Mbed TLS 3.0.0, 2.27.0 or 2.16.11 depending on the branch they're currently using.

# Prior References
https://github.com/JuliaLang/julia/issues/42634
https://github.com/JuliaLang/julia/pull/42311#issuecomment-998867938

# Potential double-free after an out of memory error

https://tls.mbed.org/tech-updates/security-advisories/mbedtls-security-advisory-2021-12

## Vulnerability
If mbedtls_ssl_set_session() or mbedtls_ssl_get_session() were to fail with MBEDTLS_ERR_SSL_ALLOC_FAILED (in an out of memory condition), then calling mbedtls_ssl_session_free() and mbedtls_ssl_free() in the usual manner would cause an internal session buffer to be freed twice, due to two structures both having valid pointers to it after a call to ssl_session_copy().

## Impact
An attacker could potentially trigger the out of memory condition, and therefore use this bug to create memory corruption, which could then be further exploited or targetted.

## Resolution
Affected users will want to upgrade to Mbed TLS 3.1.0, 2.28.0 or 2.16.12 depending on the branch they're currently using.

## Work-around
Either do not call mbedtls_ssl_session_free() (which will unfortunately cause a memory leak) or set the mbedtls_ssl_session field ticket to NULL manually, in the case where either mbedtls_ssl_set_session() or mbedtls_ssl_get_session() returns MBEDTLS_ERR_SSL_ALLOC_FAILED.

**Comment by nsslh** [comment]
Thanks for the issue and the security tag. Much appreciated.

**Comment by ViralBShah** [comment]
We need a bump on mbedtls due to the CVE for the next release. Conversation here: https://github.com/Homebrew/discussions/discussions/2749#discussioncomment-2076983

**Comment by mbauman** [comment]
So we're currently on mbedtls v2.28.0 — the latest — on [v1.8](https://github.com/JuliaLang/julia/blob/release-1.8/deps/Versions.make#L61) and above.  [v1.6](https://github.com/JuliaLang/julia/blob/release-1.7/deps/Versions.make#L63) is on v2.24.  Unfortunately, mbedtls does not follow semver and a simple backport of the upgrade is breaking.  I think our only hope of addressing these security issues on the LTS is finding and carrying the individual security patches ourselves.

**Comment by ViralBShah** [comment]
Does any Linux distro carry 2.24 in their LTS? If so, they may have backports.

**Comment by mbauman** [comment]
The only one [tracked by pkgs.org](https://pkgs.org/download/mbedtls) is [NetBSD 8.2](https://netbsd.pkgs.org/8/netbsd-amd64/mbedtls-2.24.0.tgz.html)... but it doesn't look like that's [actually the case](https://ftp.netbsd.org/pub/pkgsrc/current/pkgsrc/security/mbedtls/index.html).  In any case, they're definitely not [carrying the patches](https://github.com/NetBSD/pkgsrc/commits/trunk/security/mbedtls).

**Comment by mkitti** [comment]
We may need to admit that we are underresourced in the security domain.

To mitigate this, I suggest that we be selective about which versions we incorporate into base Julia, especially LTS releases going forward.

Perhaps we should evaluate if OS distributions have incorporated that particular version into stable or LTS releases. For example, if we were discussing 2.16 we might have an example to follow from the Debian security team.

https://packages.debian.org/search?keywords=mbedtls

Regarding mbedtls in particular, I have lost confidence that mbedtls is a good SSL/TLS dependency for Julia. Upstream's release notes clearly expect that dependents can address ABI breakage simply through recompilation. This is not the case for Julia and BinaryBuilder.

We should strongly consider switching to OpenSSL. I recognize this is easier said than done.

**Comment by mkitti** [comment]
Other options include BoringSSL
https://boringssl.googlesource.com/boringssl/ or LibreSSL https://www.libressl.org/ .

However, BoringSSL is not intended for 3rd party packages. LibreSSL has some issues regarding Linux support https://lwn.net/Articles/841664/ .

**Comment by ViralBShah** [comment]
Yes @nalimilan had proposed this as well. It is a sound strategy to use versions of mbedtls from Linux distro LTS versions. 

I know about debian/ubuntu processes but not so much how to look for the same on redhat.

**Comment by ViralBShah** [comment]
If mbedtls is breaking, can we not upgrade all the dependent stdlibs too on 1.6? Won't Pkg deal with upgrading any package dependencies?

**Comment by mkitti** [comment]
Here is an excerpt from the mbedtls release notes:

> Some fields of mbedtls_ssl_session and mbedtls_ssl_config are in a different order. This only affects applications that define such structures directly or serialize them.

https://github.com/Mbed-TLS/mbedtls/releases/tag/v2.28.0

**Comment by mkitti** [comment]
> If mbedtls is breaking, can we not upgrade all the dependent stdlibs too on 1.6? Won't Pkg deal with upgrading any package dependencies?


@giordano has answered this in the past:

https://github.com/JuliaPackaging/Yggdrasil/issues/3024
> [MbedTLS_jll](https://juliahub.com/ui/Packages/MbedTLS_jll/u5NEn/) is a mess. The soversion of at least one of the libraries provided changes between minor versions ([example 2.26 -> 2.27](https://github.com/JuliaBinaryWrappers/MbedTLS_jll.jl/commit/13717372fd1c8f6344c842fc916698a156e7dad7)), but not always (e.g., [nothing changed 2.25 -> 2.26](https://github.com/JuliaBinaryWrappers/MbedTLS_jll.jl/commit/c8c19e8baef7e5898bb516f254980ea6820af4cd)). On top of that, MbedTLS_jll is a Julia dependency, so you can't possibly build a single library that can be loaded by different Julia versions which ship different versions of MbedTLS_jll, as they'd have different ABIs:

https://github.com/JuliaPackaging/Yggdrasil/issues/4190#issuecomment-1007021777
> For point 4, downstream libraries need to do something only if the affected library changes ABI/API in the new version, which is what happens with mbedtls. But you can't just rebuild the downstream libraries against the new fixed library: there are compatibility bounds to respect, accept the fact we can't change compat bounds inside the same patch version (changing only the build number), and if the affected library is shipped by Julia, you have also to deal with julia compatibility bounds and not break existing packages

https://github.com/JuliaPackaging/Yggdrasil/pull/4179#issuecomment-1006935608
> I don't think we've ever changed binary dependencies within a minor release. Also, this will break about [1779 packages depending on MbedTLS_jll](https://juliahub.com/ui/Packages/MbedTLS_jll/u5NEn/2.27.0+0?page=2), because 2.28 breaks the ABI compared to v2.24

**Comment by mkitti** [comment]
We need to consider switching LTS to 1.7. I'm not seeing a way forward to fix these security issues in 1.6.

**Comment by giordano** [comment]
Backport the patches? As long as they don't break the ABI

**Comment by mkitti** [comment]
> Backport the patches? As long as they don't break the ABI

Yes, but who will do this? It has been many months. Do I need to moonlight as a security engineer?

**Comment by KristofferC** [comment]
Does anyone have a list of the actual commits that need to be backported?

**Comment by ViralBShah** [comment]
Just noting here that mbedtls 2.28 is LTS until 2024: https://github.com/Mbed-TLS/mbedtls/releases/tag/v2.28.0

**Comment by staticfloat** [comment]
Yggdrasil PR: https://github.com/JuliaPackaging/Yggdrasil/pull/5081

Julia binaries bump and from-source patch bump PR: https://github.com/JuliaLang/julia/pull/45848

**Comment by KristofferC** [comment]
https://tls.mbed.org has been down for me for weeks... And http://downforeveryoneorjustme.com agrees with me. Does it work for you all?

**Comment by PallHaraldsson** [comment]
Why do we need mbedTLS (or any TLS at all)?

download() is no longer implemented by Julia as of 1.6 (and in the new place, in a package, implemented by curl), and is deprecated:
https://github.com/JuliaLang/julia/blob/bb5b98e72a151c41471d8cc14cacb495d647fb7f/base/download.jl#L15

>We may need to admit that we are underresourced in the security domain. [..] I have lost confidence that mbedtls is a good SSL/TLS dependency for Julia. [..]
We should strongly consider switching to OpenSSL. I recognize this is easier said than done.

I feel bad about this since if I recall, I was the one to suggest mbedTLS (at a time when OpenSSL was a mess). I'm curious what actual programs need from mbedTLS. I see we have libssh2, any why do we need security? :) It's a serious question. It seems to me Pkg needs it, and download capability, but should none of this (including Pkg?) be bundled with Julia?

Can we drop download() in 1.8, right now, before it's released, since it's deprecated anyway, to allow getting rid of mbedTLS (or do that at the same time). Makes it easier for it to be LTS later?

>We need to consider switching LTS to 1.7.

Well, I wanted that all along (for other reasons). Since 1.7 is out of support, and if it's really considered for LTS, can download() be dropped there to for a kind of 1.7b. Since it's out of support anyway, I don't see we need to honor download() if it's resurrected.

1.7 should be compatible with 1.6 (well any new versions should be), and faster and with better newer syntax, not breaking any old. Would people be confident 1.7 has no new bugs over 1.6? We can have two LTS, at least for a while...

**Comment by ViralBShah** [comment]
Let's slow down a bit! So long as we use the LTS version of mbedtls, there is really no issue.

**Comment by staticfloat** [comment]
> https://tls.mbed.org has been down for me for weeks... And http://downforeveryoneorjustme.com agrees with me. Does it work for you all?

No, I have to use `archive.org`.

**Comment by mkitti** [comment]
> > https://tls.mbed.org has been down for me for weeks... 

It redirects to https://www.trustedfirmware.org/projects/mbed-tls/ for me.

**Comment by KristofferC** [comment]
Seems like it has been fixed now, indeed.

**Comment by mkitti** [comment]
To clarify, this is fixed on Julia 1.6.7 and Julia 1.8.

Julia 1.7 was not and will not be patched for this since it is out of support.

**Comment by mkitti** [comment]
To clarify, this is fixed on Julia 1.6.7 and Julia 1.8.

Julia 1.7 was not and will not be patched for this since it is out of support.
